### PR TITLE
Add counterexamples for parser errors

### PR DESCRIPTION
**Pull request by keltono**
This is the culmination of the work I mostly did 2 summers ago and have been finishing up since when I could.
The code is (I would say) well commented, and is mostly contained within `src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/auxiliary/counterexample`. 
The way that the code is organized and the scope of the project made this feel like a reasonable way to do things --- if need be the code could be messaged into a builder/checker/dumper, but this doesn't feel like the correct abstraction.  

After some time I decided to restrict the scope of the algorithm to only produce non-unified counterexamples, i.e examples that may not have the exact same tokens. Getting this always is (of course) undecidable in general, and the algorithm presented in the literature is massively complex slow, and very brittle to changes. I think it would be essentially unmaintainable without significant reworking of the core algorithm. 

Let me know if they're any questions or concerns with anything.
I've added Lucas and August as reviewers, let me know if I should do something else or if the review/merge process should be done in a particular way.
CC @ericvanwyk @krame505 @schwerdf .

### CODE REVIEW

File: copper-parent.iml
Comment (schwerdf) [inline copper-parent.iml:1]: Was this file included deliberately?

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompiler.java
Comment (schwerdf) [inline src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompiler.java]: The name of Graphviz is consistently misspelled across this PR. I would resolve this by removing the Graphviz references from the names altogether, since the DOT format is supported by more than one graph layout tool.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/pipeline/SpecParserParameters.java
Comment (schwerdf) [inline src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/pipeline/SpecParserParameters.java:1]: What was the reason for removing this interface?

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/skins/cup/CupSkinParser.java
Comment (schwerdf) [inline src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/skins/cup/CupSkinParser.java]: The CUP skin only supports a single grammar, so putting a grammar in the reference was not strictly necessary. However, since this is a generated file, it should be kept in sync with its source, [CupSkinGrammar.xml](https://github.com/melt-umn/copper/blob/develop/doc/specs/CupSkinGrammar.xml).

**Review by schwerdf — CHANGES_REQUESTED** [review]
Looks good in general. Specific concerns are in code comments.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/auxiliary/counterexample/CounterexampleSearch.java
Comment (krame505) [inline src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/auxiliary/counterexample/CounterexampleSearch.java:21]: I'm not going to try to review the logic here.  But nothing looks obviously wrong to me.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java
Comment (krame505) [inline src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java:205]: Would it be easy to make the default be to check whether stdout supports ANSI colors?

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java
Comment (krame505) [inline src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java]: We should probably add a flag in Silver to enable this - or make it happen with `--copperdump`.  

Also I agree with August's naming comment - should probably just be `dotOutput`, unless you are doing something that is non-standard dot and specific to GraphViz.

**Review by krame505 — COMMENTED** [review]
Looks good, but deferring to August to approve.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java
Comment (schwerdf) [reply src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java:205]: The only way I know of to check terminal capabilities like this is OS-specific (reading the `TERM` environment variable). What could be done is to default to "no color" when the output is to a log file.

Also, I see that the Javadoc in a few places refers to the ANSI codes as "ascii color codes". This should be corrected as it goes into Copper's public documentation.

Comment (keltono) [reply src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java:205]: Fixed the color code names. Working on changing the default when the output is a log file now.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java
Comment (keltono) [reply src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java]: Is there a particular preference on which of those should be done? I could add it to copperdump without too much issue, but I haven't done so yet.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/pipeline/SpecParserParameters.java
Comment (keltono) [reply src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/pipeline/SpecParserParameters.java:1]: In order to pass some of the parameters needed for the new options,  there were parts of the code that used this interface, and some that used `ParserSpecCompiler` in a way that made it impossible (or at least difficult) to pass the information down the pipeline. Collapsing it to one interface made this a lot easier and I think will make adding any flags in the future easier.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompiler.java
Comment (keltono) [reply src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompiler.java]: Makes sense, fixed.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/skins/cup/CupSkinParser.java
Comment (keltono) [reply src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/skins/cup/CupSkinParser.java]: Reverted.

**Comment by keltono** [comment]
Apologies for the slow work, I started the PR right before midterms began (my mistake!), but I (believe I) have addressed all of the issues except for not emitting color codes by default for log files, which I will finish by the end of the day. Let me know if there's anything else.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java
Comment (keltono) [reply src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java:205]: Changed, I had to add another flag to enable color to accommodate this, which I felt was better than not having an override to this behavior, or having a flag that takes a boolean as argument and needing to turn a string in to a boolean.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java
Comment (krame505) [reply src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java]: How much time does it add for a large parser, e.g. ableC?  I'd say add it to `--copperdump` if it's no more than a few seconds, otherwise make it a seperate flag.

Comment (keltono) [reply src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java]: It's pretty fast, so I'll just add it to copperdump. I did find another (more or less unrelated) bug (that I thought I squashed...) when testing that again, so I'll fix that as well when I add it. Not sure how long that will take, but hopefully I will be done by the end of the week.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompiler.java
Comment (krame505) [inline src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompiler.java]: Should be "disabled by default if using a logfile"?

**Review by krame505 — COMMENTED** [review]
Looks good to me; I'll let @schwerdf approve.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompiler.java
Comment (schwerdf) [reply src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompiler.java]: As the default log output is *stderr*, not stdout (and as the question of log colorization is not necessarily limited to the counterexamples) I would use the following text for the `-cc` and `-nc` options, which should make clear what the defaults are:
```
-cc Use ANSI color escape sequences in log output, even when logging to a file
-nc Suppress ANSI color in log output, even when logging to the terminal
```

Also, on commit 52ffd6ea, the color test seems to be the wrong way around -- I tried it out with a conflicting grammar, and the counterexamples were without color on the terminal unless I used `-cc` and with color in the logfile unless I used `-nc`.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java
Comment (schwerdf) [reply src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java:205]: We really should rework this so that the "default" color setting is the same whether invoking Copper through the CLI or the API, but we do not need to hold up the merge on this account.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/auxiliary/counterexample/CounterexampleSearch.java
Comment (schwerdf) [reply src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/compiletime/auxiliary/counterexample/CounterexampleSearch.java:21]: There are several debugging lines in here that print to `System.out`, and should be commented out.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java
Comment (keltono) [reply src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompilerParameters.java]: Alright, I looked at it and realized that I think it would require a non-trivial refactor to add to copperdump, so it would be best to add the flag to silver if it's not too much of a pain.

File: src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompiler.java
Comment (keltono) [reply src/edu.umn.cs.melt.copper.compiletime/src/main/java/edu/umn/cs/melt/copper/main/ParserCompiler.java]: I like that wording better, changed. 
Yep, looks like i put a `==` where i needed a `!=`, that also should be fixed now.

**Comment by keltono** [comment]
I think everything is ready to go for merging!

**Review by schwerdf — APPROVED** [review]

**Pull request merged**
